dmipy/utils/invariants.py

```diff
@@ -0,0 +1,576 @@
+import numpy as np
+from scipy.special import  lpmv, factorial
+from os.path import join as pjoin, dirname
+import pkg_resources
+from os.path import join
+from scipy.sparse import coo_matrix, load_npz
+from dipy.core.geometry import cart2sphere
+from dipy.reconst.shm import real_sym_sh_mrtrix
+from dipy.reconst.shm import real_sph_harm
+
+DATA_PATH = pkg_resources.resource_filename(
+    'dmipy', 'data'
+)
+
+def _get_index(l,m):
+    if l==0:
+        idx = 0
+    else:
+        i0 = ((l-1) * l)//2
+
+        idx = i0 + (m+l)
+
+    return int(idx)
+
+def _LegendreP(m,l,x):
+    return (-1)**m * lpmv(m,l,x)
+
+
+def sh_matrix(sh_order, vecs, sh_base = 'mrtrix'):
+    r""" Compute Spherical Harmonics (SH) matrix M
+    Parameters
+    ----------
+    sh_order : int, even
+        Truncation order of the SH.
+    vecs : array, Nx3
+        array unit directions [X,Y,Z].
+    sh_base : string,
+        the base of the SH.
+
+    Returns
+    --------
+    M : array,
+        The base of the SH sampled according to vecs.
+    """
+
+    r, theta, phi = cart2sphere(vecs[:, 0], vecs[:, 1], vecs[:, 2])
+    theta[np.isnan(theta)] = 0
+
+    n_c = (sh_order + 1) * (sh_order + 2) // 2
+
+    if sh_base == 'mrtrix':
+         M, m, l = real_sym_sh_mrtrix(sh_order, theta, phi)
+    else:
+        if sh_base == 'dipy':
+            shm = real_sph_harm
+        else:
+            shm = gaunt_real_sph_harm
+
+        M = np.zeros((vecs.shape[0], n_c))
+        counter = 0
+        for l in range(0, sh_order + 1, 2):
+            for m in range(-l, l + 1):
+                M[:, counter] = shm(m, l, theta, phi)
+                counter += 1
+    return M
+
+
+def _fit_sh(shell, M, L):
+    if np.all(shell==0):
+        coef = np.zeros(M.shape[1])
+    else:
+        pseudoInv = np.dot(np.linalg.inv(np.dot(M.T, M) + L), M.T)
+        coef = np.dot(pseudoInv, shell)
+    return coef
+
+def _laplace_beltrami(sh_order):
+    "Returns the Laplace-Beltrami regularisation matrix for SH basis"
+    n_c = (sh_order + 1)*(sh_order + 2)//2
+    diagL = np.zeros(n_c)
+    counter = 0
+    for l in range(0, sh_order + 1, 2):
+        for m in range(-l, l + 1):
+            diagL[counter] = (l * (l + 1)) ** 2
+            counter += 1
+
+    return np.diag(diagL)
+
+def fit_sh_to_data(data, vecs, sh_order=4, lambda_lb = 1e-03, sh_base = 'mrtrix'):
```

Comment:
I think this function needs a `mask` parameter to enable fitting not the entire volume.

When given, you can just for-loop over all masked voxels (so the different dimension cases become unnecessary), and when not given, you can default the mask to the entire volume of shape `data.shape[:-1]` and do the same thing.

I do the same thing for the `MultiCompartmentModel.fit()` functions, take a look.